cloud/ec2_util.go

```diff
@@ -610,6 +610,169 @@ func newMakeBlockDeviceMappings(mounts []MountPoint) ([]*ec2aws.BlockDeviceMappi
+	}
+	hostCost, err := onDemandCost(&pkgOnDemandPriceFetcher, os, *iType, region, dur)
+	if err != nil {
+		return 0, errors.WithStack(err)
+	}
+	return hostCost + ebsCost, nil
+}
+
+func (m *ec2Manager) costForDurationSpot(h *host.Host, start, end time.Time) (float64, error) {
+	spotDetails, err := m.client.DescribeSpotInstanceRequests(&ec2aws.DescribeSpotInstanceRequestsInput{
+		SpotInstanceRequestIds: []*string{makeStringPtr(h.Id)},
+	})
+	if err != nil {
+		return 0, err
+	}
+	instance, err := m.getInstanceInfo(*spotDetails.SpotInstanceRequests[0].InstanceId)
+	if err != nil {
+		return 0, err
+	}
+	os := osLinux
+	if strings.Contains(h.Distro.Arch, "windows") {
+		os = osWindows
+	}
+	ebsCost, err := m.blockDeviceCosts(instance.BlockDeviceMappings, end.Sub(start))
+	if err != nil {
+		return 0, errors.Wrap(err, "calculating block device costs")
+	}
+	spotCost, err := m.calculateSpotCost(instance, os, start, end)
+	if err != nil {
+		return 0, err
+	}
+	return spotCost + ebsCost, nil
+}
+
+func (m *ec2Manager) blockDeviceCosts(devices []*ec2aws.InstanceBlockDeviceMapping, dur time.Duration) (float64, error) {
+	cost := 0.0
+	if len(devices) > 0 {
+		volumeIds := []*string{}
+		for i := range devices {
+			volumeIds = append(volumeIds, devices[i].Ebs.VolumeId)
+		}
+		vols, err := m.client.DescribeVolumes(&ec2aws.DescribeVolumesInput{
+			VolumeIds: volumeIds,
+		})
+		if err != nil {
+			return 0, err
+		}
+		for _, v := range vols.Volumes {
+			// an amazon region is just the availability zone minus the final letter
+			region := azToRegion(*v.AvailabilityZone)
+			p, err := ebsCost(&pkgEBSFetcher, region, *v.Size, dur)
+			if err != nil {
+				return 0, errors.Wrapf(err, "EBS volume %v", v.VolumeId)
+			}
+			cost += p
+		}
+	}
+	return cost, nil
+}
+
+// calculateSpotCost is a helper for fetching spot price history and computing the
+// cost of a task across a host's billing cycles.
+func (m *ec2Manager) calculateSpotCost(
+	i *ec2aws.Instance, os osType, start, end time.Time) (float64, error) {
+	rates, err := m.describeHourlySpotPriceHistory(hourlySpotPriceHistoryInput{
+		iType: *i.InstanceType,
+		zone:  *i.Placement.AvailabilityZone,
+		os:    os,
+		start: *i.LaunchTime,
+		end:   end,
+	})
+	if err != nil {
+		return 0, err
+	}
+	return spotCostForRange(start, end, rates), nil
+}
+
+type hourlySpotPriceHistoryInput struct {
+	iType string
+	zone  string
+	os    osType
+	start time.Time
+	end   time.Time
+}
+
+// describeHourlySpotPriceHistory talks to Amazon to get spot price history, then
+// simplifies that history into hourly billing rates starting from the supplied
+// start time. Returns a slice of hour-separated spot prices or any errors that occur.
+func (m *ec2Manager) describeHourlySpotPriceHistory(input hourlySpotPriceHistoryInput) ([]spotRate, error) {
+	// expand times to contain the full runtime of the host
+	startFilter, endFilter := input.start.Add(-5*time.Hour), input.end.Add(time.Hour)
```

Comment:
why -5 here?

Reply:
I'm not sure. This is from the current implementation. I *think* it was to ensure that we got the whole history, but `-time.Hour` should do it, so changed.